src/main/java/org/opensearch/neuralsearch/search/HybridTopFieldDocSortCollector.java

```diff
@@ -0,0 +1,472 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.opensearch.neuralsearch.search;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+import java.util.Locale;
+import java.util.Arrays;
+import java.util.ArrayList;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import lombok.Getter;
+import lombok.Setter;
+import lombok.extern.log4j.Log4j2;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.Collector;
+import org.apache.lucene.search.Sort;
+import org.apache.lucene.search.FieldValueHitQueue;
+import org.apache.lucene.search.ScoreDoc;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.TotalHits;
+import org.apache.lucene.search.LeafCollector;
+import org.apache.lucene.search.LeafFieldComparator;
+import org.apache.lucene.search.SortField;
+import org.apache.lucene.search.TopFieldDocs;
+import org.apache.lucene.search.Scorable;
+import org.apache.lucene.search.FieldComparator;
+import org.apache.lucene.search.CollectionTerminatedException;
+import org.apache.lucene.search.FieldDoc;
+import org.apache.lucene.util.PriorityQueue;
+import org.opensearch.neuralsearch.query.HybridQueryScorer;
+import org.opensearch.common.Nullable;
+
+/*
+ Collects the TopFieldDocs after executing hybrid query. Uses HybridQueryTopDocs as DTO to handle each sub query results.
+ The individual query results are sorted as per the sort criteria sent in the search request.
+ */
+@Log4j2
+public abstract class HybridTopFieldDocSortCollector implements Collector {
+    final int numHits;
+    final HitsThresholdChecker hitsThresholdChecker;
+    int docBase;
+    LeafFieldComparator comparators[];
+    int reverseMul;
+    FieldComparator<?> firstComparator;
+    FieldValueHitQueue.Entry bottom = null;
+    FieldValueHitQueue<FieldValueHitQueue.Entry>[] compoundScores;
+    boolean queueFull[];
+    @Getter
+    private int totalHits;
+    @Getter
+    float maxScore = 0.0f;
+    int[] collectedHits;
+    @Getter
+    @Setter
+    private TotalHits.Relation totalHitsRelation = TotalHits.Relation.EQUAL_TO;
+    private Boolean searchSortPartOfIndexSort = null;
+
+    private static final TopFieldDocs EMPTY_TOPDOCS = new TopFieldDocs(
+        new TotalHits(0, TotalHits.Relation.EQUAL_TO),
+        new ScoreDoc[0],
+        new SortField[0]
+    );
+
+    // Declaring the constructor private prevents extending this class by anyone
```

Comment:
From my understanding, private constructor does not prevent extending the class. Are we enforcing this class to be a singleton